### PR TITLE
Stop SeedHostsResolver on shutdown

### PR DESCRIPTION
**Pull request by ywelsch**
Fixes an issue noticed by @DaveCTurner where tests would sometimes hang for 5 seconds when restarting a node. The reason is that the `SeedHostsResolver` is blockingly waiting on a result for the full 5 seconds when the corresponding threadpool is shut down.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/discovery/SeedHostsResolver.java [hunk 1]
```diff
@@ -98,11 +100,12 @@ public static TimeValue getResolveTimeout(Settings settings) {
                 .stream()
                 .map(hn -> (Callable<TransportAddress[]>) () -> transportService.addressesFromString(hn))
                 .collect(Collectors.toList());
-        final List<Future<TransportAddress[]>> futures;
+        final SetOnce<List<Future<TransportAddress[]>>> futures = new SetOnce<>();
         try {
-            futures = executorService.get().invokeAll(callables, resolveTimeout.nanos(), TimeUnit.NANOSECONDS);
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
+            cancellableThreads.execute(()  -> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/discovery/SeedHostsResolver.java]: Nit: no need for braces

```suggestion
            cancellableThreads.execute(() ->
```

File: server/src/main/java/org/elasticsearch/discovery/SeedHostsResolver.java [hunk 2]
```diff
@@ -98,11 +100,12 @@ public static TimeValue getResolveTimeout(Settings settings) {
...
+                futures.set(executorService.get().invokeAll(callables, resolveTimeout.nanos(), TimeUnit.NANOSECONDS));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/discovery/SeedHostsResolver.java]:
```suggestion
                futures.set(executorService.get().invokeAll(callables, resolveTimeout.nanos(), TimeUnit.NANOSECONDS)));
```

File: server/src/main/java/org/elasticsearch/discovery/SeedHostsResolver.java [hunk 3]
```diff
@@ -98,11 +100,12 @@ public static TimeValue getResolveTimeout(Settings settings) {
...
+            });
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/discovery/SeedHostsResolver.java]:
```suggestion
```

File: server/src/test/java/org/elasticsearch/discovery/SeedHostsResolverTests.java [hunk 4]
```diff
@@ -321,6 +327,62 @@ public BoundTransportAddress boundAddress() {
...
+
+            @Override
+            public TransportAddress[] addressesFromString(String address) throws UnknownHostException {
+                if ("hostname1".equals(address)) {
+                    return new TransportAddress[]{new TransportAddress(TransportAddress.META_ADDRESS, 9300)};
+                } else if ("hostname2".equals(address)) {
+                    try {
+                        conditionLatch.countDown();
+                        latch.await();
+                        return new TransportAddress[]{new TransportAddress(TransportAddress.META_ADDRESS, 9300)};
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/discovery/SeedHostsResolverTests.java]: This is unreachable, right?
```suggestion
                        throw new AssertionError("impossible");
```

**Review by DaveCTurner — APPROVED** [review]
I left a couple of nits, but LGTM either way.

File: server/src/test/java/org/elasticsearch/discovery/SeedHostsResolverTests.java (hunk 4 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/discovery/SeedHostsResolverTests.java]: yes

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/1

**Pull request merged**